Adds certificate identity field to cross-cluster API keys

gmjehovich opened the pull request:
This PR adds support for a certificate_identity field in cross-cluster API keys in order to enable linking these API keys to a certificate-based identity.

Overview of changes:

* Add certificate_identity field to API key documents and requests
* Support creation/updates of cross-cluster API keys with cert identity
* Update serialization and response classes
* Add mixed-cluster safety checks and Security Features

gmjehovich committed:
Add 'certificate_identity' field to API Key, modify create/update CrossClusterAPIKey methods to account for new field

jfreden reviewed (commented):
Good job! I took an initial look and left some comments related to the mixed cluster/upgrade scenarios. 

Looks like there are some compilation issues now and you need the new field in `TransportUpdateCrossClusterApiKeyAction` and `BulkUpdateApiKeyRequest` (pass `null` since adding the same identity to several api keys is not a use case yet).


An upgrade test in [ApiKeyBackwardsCompatibilityIT](https://github.com/elastic/elasticsearch/blob/main/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java) would be nice, to make sure that the upgrade path works as expected. `QueryableBuiltInRolesUpgradeIT` uses node features [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/QueryableBuiltInRolesUpgradeIT.java#L99) you might be able to copy/break out some of that logic for reuse. 

Since we're touching the API Key interfaces it would be good to add the `test-update-serverless` label to make sure the changes to the constructors are compatible with serverless. 

In the PKI Realm when the [username_pattern](https://www.elastic.co/docs/reference/elasticsearch/configuration-reference/security-settings) is configured, we compile the pattern as part of the setting validation. I wonder if we should do that here too? See PKIRealm [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/pki/PkiRealmSettings.java#L26-L30).

gmjehovich committed:
Add integration testing for CrossCluster API Key certificate_identity field

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Update docs/changelog/134604.yaml

gmjehovich committed:
fix merge conflicts

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Fix constructors, spotlessApply

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Fix createCrossClusterApiKey unit test

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Fix updateCrossClusterApiKeyRequestTest

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Fix CC API Key Update Message and Test Assertion

# Conflicts:
#	x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateCrossClusterApiKeyRequestTests.java

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
spotlessApply

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Fix ApiKeyBackwardsCompatibilityIT, fix update error message check

gmjehovich committed:
Fix ApiKeyBackwardsCompatibilityIT.testCertificateIdentityBackwardsCompatibility()

Commit:
[CI] Update transport version definitions

gmjehovich committed:
Remove authenticateWithApiKey from ApiKeyBackwardsCompatibilityIT.testCertificateIdentityBackwardsCompatibility UPGRADED Case

gmjehovich committed:
Add minimum version check to ApiKeyBackwardsCompatibilityIT.testCertificateIdentityBackwardsCompatibility()

Commit:
[CI] Update transport version definitions

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Add validation for certificate_identity to update path

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Fix validation logic, add capability to delete certificate_identity from API key

gmjehovich committed:
Fix NPE

Commit:
[CI] Update transport version definitions

gmjehovich committed:
Fix testing bugs that resulted from new CertificateIdentity record

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Fix ApiKeyIntegTest assertion bug

gmjehovich committed:
Merge branch 'main' into rcs_improvements

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

gmjehovich commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestUpdateCrossClusterApiKeyAction.java:
```diff
@@ -36,14 +36,16 @@ public final class RestUpdateCrossClusterApiKeyAction extends ApiKeyBaseRestHand
         PARSER.declareObject(optionalConstructorArg(), CrossClusterApiKeyRoleDescriptorBuilder.PARSER, new ParseField("access"));
         PARSER.declareObject(optionalConstructorArg(), (p, c) -> p.map(), new ParseField("metadata"));
         PARSER.declareString(optionalConstructorArg(), new ParseField("expiration"));
+        PARSER.declareString(optionalConstructorArg(), new ParseField("certificate_identity")); // Add this
```
Added omitted null vs explicit null parsing and handling

gmjehovich commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -614,6 +634,58 @@ private void createApiKeyAndIndexIt(
+    }
+
+    private static boolean isValidCertificateIdentity(String certificateIdentity) {
+        Pattern pattern = Pattern.compile("^.+$");
```
Good catch, I misunderstood how Pattern.compile worked. This pattern was overly permissive. I've updated it to just use `Pattern.compile()` now, and tests confirm that invalid regex like `[abc` are rejected as we would want

gmjehovich commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -990,6 +1074,17 @@ static XContentBuilder maybeBuildUpdatedDocument(
             );
         }
 
+        // TO DO : See if we actually want to process when this is null. Use case- when a user wants to remove
```
Changed field handling in ApiKeyService to account for explicit vs omitted null

gmjehovich commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateCrossClusterApiKeyRequest.java:
```diff
@@ -35,9 +36,9 @@ public ApiKey.Type getType() {
+        if (roleDescriptors == null && metadata == null && certificateIdentity == null) {
             validationException = addValidationError(
-                "must update either [access] or [metadata] for cross-cluster API keys",
+                "must update [access] or [metadata] or [certificate_identity] for cross-cluster API keys",
```
Ah I need to make this change too. Will do this soon. I think I will need to also change some assertions in the test files

jfreden replied:
nit: this is still unchanged.

jfreden commented on x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/ApiKeyRestIT.java:
```diff
@@ -2112,6 +2115,270 @@ public void testWorkflowsRestrictionValidation() throws IOException {
+            }""");
+        setUserForRequest(createRequest, MANAGE_SECURITY_USER, END_USER_PASSWORD);
+        final String apiKeyId = assertOKAndCreateObjectPath(client().performRequest(createRequest)).evaluate("id");
+
```
You could add an `assertNull` here to make sure `certificate_identity` is not set.

jfreden commented on x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/apikey/ApiKeyRestIT.java:
```diff
@@ -2112,6 +2115,270 @@ public void testWorkflowsRestrictionValidation() throws IOException {
+            }""", originalCertIdentity));
+        setUserForRequest(createRequest, MANAGE_SECURITY_USER, END_USER_PASSWORD);
+        final String apiKeyId = assertOKAndCreateObjectPath(client().performRequest(createRequest)).evaluate("id");
+
```
You could add an assert here to make sure it's the `originalCertIdentity` even though it's tested elsewhere.

jfreden commented on x-pack/plugin/security/src/main/java/module-info.java:
```diff
@@ -51,6 +51,7 @@
     requires org.slf4j;
     requires unboundid.ldapsdk;
     requires org.elasticsearch.logging;
+    requires jsr305;
```
What's this needed for?

gmjehovich replied:
Removed. Not sure how that ended up there, I think my IDE added it when I tried importing something I didn't end up needing.

jfreden commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:
```diff
@@ -424,4 +480,83 @@ private void assertQuery(RestClient restClient, String body, Consumer<List<Map<S
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<Boolean, RestClient> getRestClientByCertificateIdentityCapability() throws IOException {
```
The logic this shares with `RolesBackwardsCompatibilityIT` could be moved to `AbstractUpgradeTestCase`

jfreden replied:
I don't see any updates to `RolesBackwardsCompatibilityIT` or `AbstractUpgradeTestCase` was it not possible to put that in shared code? No worries if it doesn't make sense :)

jfreden commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java:
```diff
@@ -3194,14 +3203,612 @@ public void testValidateOwnerUserRoleDescriptorsWithWorkflowsRestriction() {
         assertThat(e2.getMessage(), containsString("owner user role descriptors must not include restriction"));
     }
 
+    public void testApiKeyDocSerializationWithCertificateIdentity() throws IOException {
```
The coverage of these tests is great. I wonder if the tests across `ApiKeyRestIT`, `ApiKeyServiceTests` and `ApiKeyIntegTests` could be consolidated and maybe more focused? We're testing the same thing a couple of times. Not sure exactly how that would be divided but happy to help out over zoom.

jfreden reviewed (commented):
Great job with this! I've left some comments.

gmjehovich commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/UpdateApiKeyRequestTranslator.java:
```diff
@@ -59,9 +61,20 @@ public UpdateApiKeyRequest translate(RestRequest request) throws IOException {
+                payload.roleDescriptors,
+                payload.metadata,
+                payload.expiration,
+                payload.certificateIdentity
```
Now that I am looking into it, I am not sure UpdateApiKeyRequest should even be able to accept `certificate_identity` in its constructor or have it as a field since it's only used for REST keys. I think I just had it added initially out of confusion

gmjehovich committed:
Remove certificate_identity field from UpdateApiKeyRequestTranslator, UpdateApiKeyRequest class. Add testing

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Clean up certificate_identity testing in ApiKeyServiceTests

gmjehovich committed:
Delete redundant integ tests

gmjehovich commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/ApiKeyBackwardsCompatibilityIT.java:
```diff
@@ -206,6 +206,62 @@ public void testCreatingAndUpdatingApiKeys() throws Exception {
+    public void testCertificateIdentityBackwardsCompatibility() throws Exception {
+        assumeTrue(
+            "certificate identity backwards compatibility only relevant when upgrading from pre-9.2.0",
+            UPGRADE_FROM_VERSION.before(Version.V_9_2_0)
```
Is this the correct version, or is this feature going to be added in 9.3.0?

gmjehovich committed:
Consolidate redundant code in ApiKeyBackwardsCompatibilityIT

jfreden commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CertificateIdentity.java:
```diff
@@ -0,0 +1,29 @@
+                Pattern.compile(value);
+            } catch (PatternSyntaxException e) {
+                throw new IllegalArgumentException(
+                    "Invalid certificate_identity format: [" + value + "]. Must be a valid regex name pattern.",
```
```suggestion
                    "Invalid certificate_identity format: [" + value + "]. Must be a valid regex.",
```

jfreden commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java:
```diff
@@ -3202,6 +3208,123 @@ public void testValidateOwnerUserRoleDescriptorsWithWorkflowsRestriction() {
         assertThat(e2.getMessage(), containsString("owner user role descriptors must not include restriction"));
     }
 
+    public void testMaybeBuildUpdatedDocumentCertificateIdentityHandling() throws Exception {
```
Nice!

jfreden reviewed (approved):
LGTM! 

Great job on this! I only have a couple of non-blocking comments. Let me know if you want to discuss more in detail.

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Merge branch 'main' into rcs_improvements

gmjehovich committed:
Change cert_identity validation failure message

gmjehovich committed:
Update cert_identity version to 9.3.0 in APIKey BWC Test

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
CertID parser no longer differentiates between explicit vs implicit null

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Move shared bwc test logic to AbstractUpgradeTestCase

Commit:
[CI] Auto commit changes from spotless

gmjehovich committed:
Rename cleanUp method in TokenBackwwardsCompatbilityIT

gmjehovich committed:
Merge branch 'main' into rcs_improvements
